File: JavaScript/Queue.js
// FIFO data structure (First In, First Out). 
// First element to come in will be first to come out.

class Node {
    constructor(val, next=null) {
        this.val = val;
        this.next = next;
    }
}

class Queue {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    push(itm) {
        let newNode = new Node(itm);
        if (!this.head) {
            this.head = newNode;
            this.tail = this.head;
        }
        else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
    }
    pop() {
        if (!this.head) return undefined;
        let removedItem = this.head.val;
        this.head = this.head.next;
        this.length--;
        return removedItem;
    }
    print() {
        if (!this.head) return undefined;
        let current = this.head;
        let i = 0;
        while (current) {
            console.log(`[${i}]:`, current.val);
            current = current.next;
            i++;
        }
    }
}

// test drive code
let que = new Queue();
for (let i = 0; i < 5; i++) {
    que.push("a" + i);
}

console.log(que.pop());
console.log(que.pop());
console.log(que.pop());
que.print();

File: GoLang/model/queue.go
package model

import "errors"

var errQueueEmpty = errors.New("queue is empty")

type Queue struct {
	Head *SIntNode
	Tail *SIntNode
	Size int
}

func (q *Queue) Add(x int) {
	node := &SIntNode{Value: x}
	if q.Size == 0 {
		q.Head = node
		q.Tail = node
	} else {
		q.Tail.Next = node
	}
	q.Size++
}

func (q *Queue) Remove() (int, error) {
	if q.Size == 0 {
		return 0, errQueueEmpty
	}
	node := q.Head
	q.Head = q.Head.Next
	q.Size--
	return node.Value, nil
}
